[OG-112] Exception handling

LisieckaK00 opened the pull request:
**Key takeaways:**
- Frontend handles backend errors globally by displaying a toast, so there's no need to manually add `onError` to `useQuery` or use `error` in `toast.promise()`.
- Avoid using raw `fetch`, use the methods provided by `ApiClient`.
- Use `ErrorComponent` for rendering error states. It includes two variants: one for full-page errors and a compact version suitable for smaller spaces, such as three-column layouts or modals.
- Backend should return error messages in Polish, as they are displayed directly on the frontend.
---

**General description of changes.**

Backend:
- Changed all error messages to Polish, so they can be displayed directly on the frontend.
- Switched to using `ProblemDetail` as the standard error response. This requires adding an extra property in `application.properties` for proper configuration

Frontend:
- Added an `ApiError` class that is thrown whenever the backend returns an error. `ApiError` messages are always displayed as toasts — this is handled globally.
- Created an `ErrorComponent` that displays a title and a message. I also added basic styles for it
- Added dedicated methods per API endpoint

krudny commented:
@kligarski was first with fetch wrapper, but both of them are some kind of incomplete. In my mind there was a generic class/object with 4 methods (get, post, put, delete) . We are getting rid of passing params every fetch and we could for example call apiFetch.delete(url) from service layer.

Also fact with hook logout/401 I discussed with @kligarski should be solved. 

Please make one unified solution to both tasks, I was saying that they are related with each other, I dont know why 2 person have those 2 tasks. 

I know it is a draft, but just for your knowledge, handling errors inside rendered components is also a part of this task.

krudny commented:
If its toast then probably there's a reason for that. If not, then something unexpected happened. And in both situations we should handle that gracefully not leaving user with blank page.

LisieckaK00 commented on polymorphia-frontend/hooks/course/useSubmissionsUpdate/index.tsx:
```diff
@@ -20,7 +20,7 @@ export default function useSubmissionsUpdate({
     mutationFn: (submissionDetails: SubmissionDetailsResponseDTO) => {
       if (!target) {
-        throw new Error("Wystąpił błąd podczas aktualizacji oddanych zadań.");
+        throw new Error("Wystąpił błąd podczas aktualizacji oddanych zadań."); // TODO: ?
```
just noticed this, will do sth about it

krudny commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/markdown/MarkdownType.java:
```diff
@@ -29,6 +29,6 @@ public static MarkdownType fromValue(String value) {
             }
         }
-        throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Markdown type not found");
+        throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Nie znaleziono typu Markdown.");
```
markdown lowercase

krudny commented on polymorphia-frontend/components/error-state/index.css:
```diff
@@ -0,0 +1,14 @@
+@import "../../styles/globals.css";
+
+.error-state {
+  @apply flex flex-col items-center justify-center gap-4 p-8 text-center m-auto w-full max-w-xl;
```
I am saying this 150 time. Every pr except @kligarski has this comment. We are using `flex-centered` and `flex-col-centered` utility classes defined in `globals.css`

LisieckaK00 replied:
Thank you for pointing that out. I wasn’t aware of those utility classes since it’s been quite a while since I last worked directly with CSS. I’ve just updated the code to use `flex-col-centered`.

LisieckaK00 commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/exception/GeneralExceptionHandler.java:
```diff
@@ -1,54 +1,85 @@
+    @ExceptionHandler({MethodArgumentTypeMismatchException.class, ConstraintViolationException.class, InvalidTypeIdException.class})
+    public ResponseEntity<ProblemDetail> handleArgumentNotValidException(Exception ex) {
+        log.error(ex.getMessage());
+        return getResponseEntity(HttpStatus.BAD_REQUEST, "Niepoprawny typ parametru.");
```
We may just leave logging the error as we have it here, but return some more general message

kligarski replied:
It looks reasonable here.

krudny commented on polymorphia-frontend/services/api/client.ts:
```diff
@@ -0,0 +1,69 @@
+    method: "POST",
+    headers:
+      headers ??
+      (addJsonContentType ? { "Content-Type": "application/json" } : undefined),
```
Not sure but this seems to be completely invalid. I dont know why there is ?? operator. If someone passess headers then addJson flag will be skipped.

krudny commented on polymorphia-frontend/services/api/error.ts:
```diff
@@ -0,0 +1,6 @@
+    super(message);
+    this.name = "ApiError";
+  }
+}
```
Not sure, but does this contain http status code?

krudny commented on polymorphia-frontend/services/api/client.ts:
```diff
@@ -0,0 +1,69 @@
+
+export const putEndpoint = (
+  path: string,
+  body?: BodyInit,
```
We are not using all possible bodies, we use only form url encoded and json so this type is way to large. 

Then answering question, by narrowing type to just those two, you can make just one Json.stringify or .toString() depending on body type inside post method

krudny commented on polymorphia-frontend/app/(logged-in)/knowledge-base/chests/page.tsx:
```diff
@@ -20,11 +21,16 @@ export default function Chests() {
 
   if (error) {
-    return <div>Error loading chests: {error.message}</div>;
+    return <ErrorState message="Nie udało się załadować skrzynek." />;
```
Sometimes we use `error`, sometimes `isError`.

There you are skipping error values, so `isError` is sufficient. 

Not sure about this double conditions. 
We could merge into one condition with || (just one should be true and we are sure that we have to show ErrorState), and depending on chest.length display proper error message. 

Less code == better

LisieckaK00 replied:
I changed it everywhere to `isError`. 

I don't know if I understand the reasoning for one condition. Now we have:
```js
if (isError) {
  return <ErrorComponent message="Nie udało się załadować skrzynek." />;
}

if (!chests || chests.length === 0) {
  return (
    <ErrorComponent
      title="Brak danych"
      message="Skrzynki nie zostały zdefiniowane."
    />
  );
}
```
But when doing one confition with `||` we need to do second `if` anyway to display proper message:
```js
if ((isError) || (!chests || chests.length === 0)) {
  if (!chests || chests.length === 0) {
    return (
        <ErrorComponent
            title="Brak danych"
            message="Skrzynki nie zostały zdefiniowane."
        />
    );
  }
  return <ErrorComponent message="Nie udało się załadować skrzynek." />;
}
```
So I don't understant what you mean by less code here, or maybe I didn't understand your comment

krudny commented:
I got also english "no static resources" toast.

krudny commented:
Some errors have . at the end, some not, maybe we could standarize that.

krudny commented:
<img width="717" height="626" alt="image" src="https://github.com/user-attachments/assets/f029cef7-4442-4fdc-a1bc-d0e08e0fbde6" />

same error doubled, any of them doesn't provide reason.

(account already exists -> right now if any account of csv exists we throw error and block whole process. I dont know, but it might be good to throw which account is already created. But thats just one error, but what if we have 10 existing accounts? 10 toast at one time, or 1 toast every submit try? please ask @kligarski for opinion).

krudny commented:
<img width="1065" height="850" alt="image" src="https://github.com/user-attachments/assets/8f74d07f-d360-44ac-9efa-d866de45b02a" />

it was error there, now its toast and infinite loading. toast is good, infinite loading is bad

krudny commented:
probably out of scope, but i am able to crash opening chest system, button is not locked after first request, then i dont know what is happening 

<img width="407" height="813" alt="image" src="https://github.com/user-attachments/assets/761b677a-d933-44c9-ae9c-91e2bc2373c5" />

mm 

<img width="403" height="910" alt="image" src="https://github.com/user-attachments/assets/e80eca17-35e5-4fd3-acab-6f49ecf65a81" />


not your part, will forward to @ustankie

LisieckaK00 commented:
I won’t address all comments just yet.
I started by modifying the logic for the four HTTP methods. I uploaded two files: `/services/api/client.ts` and `/services/api/types.ts`. An example of how calling the methods looks now:
```ts
await ApiClient.post<HallOfFameResponseDTO>("/hall-of-fame", {
  courseId: courseId,
  page: page,
  size: size,
  searchTerm: !searchTerm || searchTerm.trim() === "" ? "" : searchTerm,
  searchBy: searchBy,
  sortBy: sortBy,
  sortOrder: sortOrder,
  groups: groups && !groups?.includes("all") ? groups : [],
})
```
However, I ran into two issues and would like your opinions (@krudny, @kligarski):
1. `export type ApiBody = string | FormData | URLSearchParams | object;`
Using `object` doesn’t really restrict the type, and it doesn’t feel like a good solution. The problem is that I’m not sure how else to allow passing arbitrary DTOs as the request body so that we can `JSON.stringify()` them inside the method. @krudny, you mentioned wanting to restrict the types, but I think this approach doesn’t really achieve that.
An alternative would be to require already stringified bodies (`JSON.stringify()` done before calling the method).
I don’t see a better option right now, so I’d appreciate it if you could either pick one of these two approaches or suggest another idea.
2. I’m not sure if I understood the point about `fetchJson` correctly. I tried doing this inside the `request` method: `return (await response.json()) as TResponse`.  The issue is that sometimes we don’t actually care about the response body, but we still need to specify a type. We can pass `unknown` or `never`, but that feels a bit awkward. An alternative would be just return `Response`, but that would mean handling parsing manually in multiple places.
So I’d like to know which approach you’d prefer, or if you have a better suggestion.
This is how it looks now:
```ts
inviteUser: async (request: InviteRequestDTO): Promise<void> => {
  await ApiClient.post<unknown>("/invitation/course", request);
},
```

krudny commented:
1. Structure of objects we pass into JSON.Stringify() is fixed and possible to typesafe. Keys are always strings, values are either string, number, undefined, null or boolean. This is "primitive" json values. Also value could be another json value object. Also it is possible to have array of json value objects eg:

```ts
{
  "sample key": ["sample value", "sample value"]
}
```

moreover, recursive json values are also supported

```ts
{
  "key": {
    "another key": value
  }
}
```

Probably that can solve issue 1

kligarski commented on polymorphia-frontend/services/api/client.ts:
```diff
@@ -0,0 +1,87 @@
+  if (!headers && !hasJsonBody) {
+    return undefined;
+  }
+  const result = new Headers(headers ?? {});
```
```suggestion
  const result = new Headers(headers);
```
With the condition above it's not necessary.

kligarski commented on polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/student/AnimalService.java:
```diff
@@ -76,14 +72,14 @@ public void createAnimal(CreateAnimalRequestDto requestDTO) {
             assignment.setAnimal(savedAnimal);
         } catch (Exception e) {
-            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, FAILED_TO_CREATE_ANIMAL);
+            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Nie udało się utowrzyć zwierzaka.");
```
```suggestion
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Nie udało się utworzyć zwierzaka.");
```

kligarski reviewed (commented):
> <img alt="Screenshot 2025-11-03 at 18 23 01" width="825" height="646" src="https://private-user-images.githubusercontent.com/134285490/509085939-33da2865-be79-4b16-b5c5-712dc866004e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjIyODA1NTMsIm5iZiI6MTc2MjI4MDI1MywicGF0aCI6Ii8xMzQyODU0OTAvNTA5MDg1OTM5LTMzZGEyODY1LWJlNzktNGIxNi1iNWM1LTcxMmRjODY2MDA0ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMTA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTEwNFQxODE3MzNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MjkyMmJhZjljZTRiMDU4MDg1OTQ5M2Q2ODliN2U4ZTAxMTM0MmUxYWMyZDAyMzhkZWYwOTM5Yzc1ZGVmOTMzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.o18MybqcIhzSbLxxBVCMUkBnwUZj0YJlibYcvuCMw2I">
> 
> same error doubled, any of them doesn't provide reason.
> 
> (account already exists -> right now if any account of csv exists we throw error and block whole process. I dont know, but it might be good to throw which account is already created. But thats just one error, but what if we have 10 existing accounts? 10 toast at one time, or 1 toast every submit try? please ask @kligarski for opinion).

This would benefit from custom error handling, e.g. students that already exist can have red border/red background/some icon. This is obviously out of scope of this PR. For now maybe just stick to one toast.

---

I'll wait with further frontend review as ApiClient isn't used yet.

LisieckaK00 commented:
**Note:** I'll be marking comments that I've addressed by thumbs up, so that I don't get lost. 

> outdated branch / oops page

Next thing I'll be doing, after I push recent after review changes

> same error doubled, any of them doesn't provide reason, csv handling

Doubled bc I missed frontend handling in this case, as it was in a different file, removed it now. When it comes to changing the logic of handling CSV I think it is not a scope of this PR. We may create an issue / jira ticket

> Some errors have . at the end, some not, maybe we could standarize that.

As said above. On backend messages should be in the same style now, the inconsistency came from the message on frontend I missed.

> hof Infinite loading

Updated, now there is an error page

> no auorized / tbh this makes no sense, user now cant do anything

We handle unauthorized now so it shouldn't be an issue anymore.

> I got also english "no static resources" toast.

How did you do it? I can't find the message in the code and never saw that toast, so I don't know how to reproduce it

LisieckaK00 commented on polymorphia-frontend/services/api/client.ts:
```diff
@@ -0,0 +1,87 @@
+    throw new ApiError(await readErrorMessage(response));
+  }
+
+  return (await response.json()) as TResponse;
```
I updated the logic, so now backend returns 204 co content. The `/login` returns 200 with empty body, so there is a `content-length` check.

We don't serialize returned objects on our own on the backend. Do you think that a check for incorrect json syntax is needed anyway?

kligarski commented on polymorphia-frontend/services/api/client.ts:
```diff
@@ -67,21 +67,60 @@ async function request<TResponse>({
+  if (
+    response.status === 204 ||
+    (contentLength && Number(contentLength) === 0)
+  ) {
```
We should also include a case when contentLength is null, right?

<img width="940" height="567" alt="Image" src="https://github.com/user-attachments/assets/4a590b3e-e2ef-4811-8813-fe399690b9b4" />

LisieckaK00 replied:
We don't expect it to happen - right now the only issue is with `/login` that returns 200 with no content. But the header is set. But if you think it'd be better to add anyway, I'll try to find some way to do it

krudny commented on polymorphia-frontend/providers/user/index.tsx:
```diff
@@ -2,14 +2,21 @@ import { createContext, ReactNode } from "react";
+      <NotFound
+        actionLabel={"Strona główna"}
+        subtitle={"Coś poszło nie tak."}
+        rerouteTo={"/login"}
```
{} are not neccesary

krudny commented:
<img width="787" height="210" alt="image" src="https://github.com/user-attachments/assets/4513b1d4-8474-44be-b921-c241e04bf74f" />

krudny commented:
Theres infinite loop after entering invalid url, then "powrót" (to invalid url again). Maybe default path for role (/profile, /course-group), maybe double .back(), maybe redirect to /

krudny commented:
Backend not working

https://github.com/user-attachments/assets/a374dd1d-f689-44d8-a640-da458c997443

krudny commented:
After redirecting user to homepage via ooops i think that we dont logout user in backend security context.

krudny commented:
I asked you to make a small in change manual invite modal, i see it is not changed

LisieckaK00 commented on polymorphia-frontend/services/csv/index.ts:
```diff
@@ -1,9 +1,9 @@
+import { ApiClient } from "@/services/api/client";
 
 const CSVService = {
   getCSVHeaders: async (
```
moved
